File: public/books/fraud-detection/code/03_logistic_fscore.R
# =====================================================
# code/03_logistic_fscore.R
# 第 3 章：逻辑回归 + Dechow F-Score
# Model A：42 个特征全集（28 原始 + 14 衍生比率）
# Model B：Dechow et al. 2011 Table 7 Model 1 的 7 变量
# 输出：测试集 AUC / NDCG@100 / Recall@1% / Precision@1%
#       Enron (gvkey=6127) 与 Tyco (gvkey=10787) 在 fyear=2000 的打分
# =====================================================

suppressPackageStartupMessages({
  library(tidyverse)
  library(here)
  library(pROC)
})

set.seed(2026)

# ---------- 数据加载 ----------
d <- read_csv(here::here("data", "bao2020_full.csv"),
              show_col_types = FALSE)

cat(sprintf("原始数据: %d firm-years\n", nrow(d)))

# 28 原始 Compustat + 14 衍生比率 = 42 特征
raw_vars <- c("act","ap","at","ceq","che","cogs","csho","dlc","dltis","dltt",
              "dp","ib","invt","ivao","ivst","lct","lt","ni","ppegt","pstk",
              "re","rect","sale","sstk","txp","txt","xint","prcc_f")
ratio_vars <- c("dch_wc","ch_rsst","dch_rec","dch_inv","soft_assets",
                "ch_cs","ch_cm","ch_roa","issue","bm","dpi","reoa","EBIT","ch_fcf")
all_vars  <- c(raw_vars, ratio_vars)

# Dechow 2011 Table 7 Model 1 的 7 个变量
dechow_vars <- c("ch_rsst","dch_rec","dch_inv","soft_assets",
                 "ch_cs","ch_roa","issue")

# ---------- 切分前先过滤可用样本 ----------
d_full <- d %>% drop_na(all_of(all_vars))
d_dech <- d %>% drop_na(all_of(dechow_vars))

cat(sprintf("Model A 可用 (42 特征无 NA): %d 行 (丢 %d)\n",
            nrow(d_full), nrow(d) - nrow(d_full)))
cat(sprintf("Model B 可用 (7 特征无 NA): %d 行 (丢 %d)\n",
            nrow(d_dech), nrow(d) - nrow(d_dech)))

split_set <- function(df) {
  list(
    train = df %>% filter(fyear >= 1991, fyear <= 2002),
    valid = df %>% filter(fyear >= 2003, fyear <= 2008),
    test  = df %>% filter(fyear >= 2009, fyear <= 2014)
  )
}

sA <- split_set(d_full)
sB <- split_set(d_dech)

cat(sprintf("Model A 切分: train=%d (fraud=%d), valid=%d (fraud=%d), test=%d (fraud=%d)\n",
            nrow(sA$train), sum(sA$train$misstate),
            nrow(sA$valid), sum(sA$valid$misstate),
            nrow(sA$test),  sum(sA$test$misstate)))
cat(sprintf("Model B 切分: train=%d (fraud=%d), valid=%d (fraud=%d), test=%d (fraud=%d)\n",
            nrow(sB$train), sum(sB$train$misstate),
            nrow(sB$valid), sum(sB$valid$misstate),
            nrow(sB$test),  sum(sB$test$misstate)))

# ---------- 评估函数 ----------
eval_metrics <- function(scores, labels, k = 100, p = 0.01) {
  ord <- order(-scores)
  s_ord <- scores[ord]; y_ord <- labels[ord]
  n <- length(labels); n_pos <- sum(labels)
  # AUC
  auc <- as.numeric(pROC::auc(pROC::roc(labels, scores,
                                        quiet = TRUE, direction = "<")))
  # NDCG@k
  rel <- y_ord[seq_len(min(k, n))]
  dcg <- sum(rel / log2(seq_along(rel) + 1))
  ideal <- c(rep(1, min(k, n_pos)), rep(0, max(k - n_pos, 0)))
  idcg <- sum(ideal / log2(seq_along(ideal) + 1))
  ndcg <- if (idcg > 0) dcg / idcg else 0
  # Recall@p / Precision@p
  k_p <- ceiling(n * p)
  hit <- sum(y_ord[seq_len(k_p)])
  recall_p <- hit / n_pos
  prec_p   <- hit / k_p
  list(AUC = auc, NDCG = ndcg, RecallP = recall_p, PrecP = prec_p,
       k_p = k_p, hit = hit, n = n, n_pos = n_pos)
}

# ---------- Model A: 42 特征全集 ----------
cat("\n========== Model A: 42 特征全集 ==========\n")
form_A <- as.formula(paste("misstate ~", paste(all_vars, collapse = " + ")))
mA <- glm(form_A, data = sA$train, family = binomial)

predA_test <- predict(mA, newdata = sA$test, type = "response")
metA <- eval_metrics(predA_test, sA$test$misstate)
cat(sprintf("Model A 测试集: AUC=%.4f  NDCG@100=%.4f  Recall@1%%=%.4f  Precision@1%%=%.4f\n",
            metA$AUC, metA$NDCG, metA$RecallP, metA$PrecP))
cat(sprintf("  测试 n=%d, 阳性=%d, 1%% 名额=%d, 命中=%d\n",
            metA$n, metA$n_pos, metA$k_p, metA$hit))

# ---------- Model B: Dechow 7 变量 ----------
cat("\n========== Model B: Dechow F-Score 7 变量 ==========\n")
form_B <- as.formula(paste("misstate ~", paste(dechow_vars, collapse = " + ")))
mB <- glm(form_B, data = sB$train, family = binomial)

cat("\n--- Dechow 7 变量系数 ---\n")
coef_tbl <- summary(mB)$coefficients
print(round(coef_tbl, 6))

predB_test <- predict(mB, newdata = sB$test, type = "response")
metB <- eval_metrics(predB_test, sB$test$misstate)
cat(sprintf("\nModel B 测试集: AUC=%.4f  NDCG@100=%.4f  Recall@1%%=%.4f  Precision@1%%=%.4f\n",
            metB$AUC, metB$NDCG, metB$RecallP, metB$PrecP))
cat(sprintf("  测试 n=%d, 阳性=%d, 1%% 名额=%d, 命中=%d\n",
            metB$n, metB$n_pos, metB$k_p, metB$hit))

# F-Score = 预测概率 / 无条件舞弊率
uncond_rate <- mean(sB$train$misstate)
cat(sprintf("\nModel B 训练集无条件舞弊率: %.6f\n", uncond_rate))

# ---------- 案例公司：Enron 2000 / Tyco 2000 ----------
cat("\n========== 案例公司 fyear=2000 打分 ==========\n")

case_score <- function(model, df_split_test, src_df, gvkey_v, fyear_v) {
  row <- src_df %>% filter(gvkey == gvkey_v, fyear == fyear_v)
  if (nrow(row) == 0) return(NA_real_)
  predict(model, newdata = row, type = "response")
}

# 用切分前的"可用样本"取案例公司行
enron_A <- case_score(mA, NULL, d_full,  6127, 2000)
tyco_A  <- case_score(mA, NULL, d_full, 10787, 2000)
enron_B <- case_score(mB, NULL, d_dech,  6127, 2000)
tyco_B  <- case_score(mB, NULL, d_dech, 10787, 2000)

cat(sprintf("Enron 2000 (gvkey=6127):  Model A p=%.4f, Model B p=%.4f, F-Score=%.2f\n",
            enron_A, enron_B, enron_B / uncond_rate))
cat(sprintf("Tyco  2000 (gvkey=10787): Model A p=%.4f, Model B p=%.4f, F-Score=%.2f\n",
            tyco_A,  tyco_B,  tyco_B  / uncond_rate))

# ---------- 测试期排名分位 ----------
rank_pos <- function(scores, target) {
  ord <- order(-scores)
  which(ord == target)
}

cat("\n========== 写入汇总 ==========\n")
cat(sprintf("[A] AUC=%.4f  NDCG=%.4f  R@1%%=%.4f  P@1%%=%.4f\n",
            metA$AUC, metA$NDCG, metA$RecallP, metA$PrecP))
cat(sprintf("[B] AUC=%.4f  NDCG=%.4f  R@1%%=%.4f  P@1%%=%.4f\n",
            metB$AUC, metB$NDCG, metB$RecallP, metB$PrecP))
cat("Done.\n")

File: public/books/fraud-detection/code/03_logistic_fscore.py
# =====================================================
# code/03_logistic_fscore.py
# 第 3 章：逻辑回归 + Dechow F-Score (Python 版)
# Model A：42 特征全集（28 原始 + 14 衍生比率）
# Model B：Dechow et al. 2011 Table 7 Model 1 的 7 变量
# =====================================================

import os, random
import numpy as np
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score

random.seed(2026)
np.random.seed(2026)

HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.dirname(HERE)
csv_path = os.path.join(ROOT, "data", "bao2020_full.csv")

d = pd.read_csv(csv_path)
print(f"原始数据: {len(d)} firm-years")

raw_vars = ["act","ap","at","ceq","che","cogs","csho","dlc","dltis","dltt",
            "dp","ib","invt","ivao","ivst","lct","lt","ni","ppegt","pstk",
            "re","rect","sale","sstk","txp","txt","xint","prcc_f"]
ratio_vars = ["dch_wc","ch_rsst","dch_rec","dch_inv","soft_assets",
              "ch_cs","ch_cm","ch_roa","issue","bm","dpi","reoa","EBIT","ch_fcf"]
all_vars = raw_vars + ratio_vars
dechow_vars = ["ch_rsst","dch_rec","dch_inv","soft_assets",
               "ch_cs","ch_roa","issue"]

d_full = d.dropna(subset=all_vars).copy()
d_dech = d.dropna(subset=dechow_vars).copy()

print(f"Model A 可用 (42 特征无 NA): {len(d_full)} 行 (丢 {len(d) - len(d_full)})")
print(f"Model B 可用 (7 特征无 NA):  {len(d_dech)} 行 (丢 {len(d) - len(d_dech)})")


def split_set(df):
    return {
        "train": df[(df["fyear"] >= 1991) & (df["fyear"] <= 2002)],
        "valid": df[(df["fyear"] >= 2003) & (df["fyear"] <= 2008)],
        "test":  df[(df["fyear"] >= 2009) & (df["fyear"] <= 2014)],
    }


sA = split_set(d_full)
sB = split_set(d_dech)

for tag, s in (("A", sA), ("B", sB)):
    print(f"Model {tag} 切分: "
          f"train={len(s['train'])} (fraud={int(s['train']['misstate'].sum())}), "
          f"valid={len(s['valid'])} (fraud={int(s['valid']['misstate'].sum())}), "
          f"test={len(s['test'])} (fraud={int(s['test']['misstate'].sum())})")


def eval_metrics(scores, labels, k=100, p=0.01):
    scores = np.asarray(scores); labels = np.asarray(labels)
    n = len(labels); n_pos = int(labels.sum())
    auc = roc_auc_score(labels, scores)
    order = np.argsort(-scores)
    y_ord = labels[order]
    rel = y_ord[:min(k, n)]
    dcg = float(np.sum(rel / np.log2(np.arange(1, len(rel) + 1) + 1)))
    ideal = np.concatenate([np.ones(min(k, n_pos)),
                            np.zeros(max(k - n_pos, 0))])
    idcg = float(np.sum(ideal / np.log2(np.arange(1, len(ideal) + 1) + 1)))
    ndcg = dcg / idcg if idcg > 0 else 0.0
    k_p = int(np.ceil(n * p))
    hit = int(y_ord[:k_p].sum())
    recall_p = hit / n_pos
    prec_p = hit / k_p
    return dict(AUC=auc, NDCG=ndcg, RecallP=recall_p, PrecP=prec_p,
                k_p=k_p, hit=hit, n=n, n_pos=n_pos)


def fit_logit(train, vars_, C=1e8, class_weight=None):
    # C 极大 ≈ 无正则，逼近 R glm 的最大似然解
    # newton-cg 在 42 特征量纲差异大时比 lbfgs 更稳
    X = train[vars_].values
    y = train["misstate"].values.astype(int)
    clf = LogisticRegression(
        penalty="l2", C=C, solver="newton-cg",
        max_iter=2000, class_weight=class_weight,
        random_state=2026, tol=1e-7,
    )
    clf.fit(X, y)
    return clf


# ---------- Model A ----------
print("\n========== Model A: 42 特征全集 ==========")
mA = fit_logit(sA["train"], all_vars)
predA_test = mA.predict_proba(sA["test"][all_vars].values)[:, 1]
metA = eval_metrics(predA_test, sA["test"]["misstate"].values)
print(f"Model A 测试集: AUC={metA['AUC']:.4f}  NDCG@100={metA['NDCG']:.4f}  "
      f"Recall@1%={metA['RecallP']:.4f}  Precision@1%={metA['PrecP']:.4f}")
print(f"  测试 n={metA['n']}, 阳性={metA['n_pos']}, "
      f"1% 名额={metA['k_p']}, 命中={metA['hit']}")

# ---------- Model B ----------
print("\n========== Model B: Dechow F-Score 7 变量 ==========")
mB = fit_logit(sB["train"], dechow_vars)
print("--- Dechow 7 变量系数 ---")
print(f"  (Intercept) {mB.intercept_[0]: .6f}")
for name, b in zip(dechow_vars, mB.coef_[0]):
    print(f"  {name:<12s} {b: .6f}")

predB_test = mB.predict_proba(sB["test"][dechow_vars].values)[:, 1]
metB = eval_metrics(predB_test, sB["test"]["misstate"].values)
print(f"\nModel B 测试集: AUC={metB['AUC']:.4f}  NDCG@100={metB['NDCG']:.4f}  "
      f"Recall@1%={metB['RecallP']:.4f}  Precision@1%={metB['PrecP']:.4f}")
print(f"  测试 n={metB['n']}, 阳性={metB['n_pos']}, "
      f"1% 名额={metB['k_p']}, 命中={metB['hit']}")

uncond = float(sB["train"]["misstate"].mean())
print(f"\nModel B 训练集无条件舞弊率: {uncond:.6f}")

# ---------- 案例公司 ----------
def case_pred(model, df, vars_, gvkey_v, fyear_v):
    row = df[(df["gvkey"] == gvkey_v) & (df["fyear"] == fyear_v)]
    if len(row) == 0:
        return None
    return float(model.predict_proba(row[vars_].values)[0, 1])

print("\n========== 案例公司 fyear=2000 打分 ==========")
enron_A = case_pred(mA, d_full, all_vars,  6127, 2000)
tyco_A  = case_pred(mA, d_full, all_vars, 10787, 2000)
enron_B = case_pred(mB, d_dech, dechow_vars,  6127, 2000)
tyco_B  = case_pred(mB, d_dech, dechow_vars, 10787, 2000)
print(f"Enron 2000 (gvkey=6127):  Model A p={enron_A:.4f}, "
      f"Model B p={enron_B:.4f}, F-Score={enron_B/uncond:.2f}")
print(f"Tyco  2000 (gvkey=10787): Model A p={tyco_A:.4f}, "
      f"Model B p={tyco_B:.4f}, F-Score={tyco_B/uncond:.2f}")

# ---------- 备选：class_weight='balanced' 提示 ----------
print("\n[备选] class_weight='balanced' 在不平衡数据上是常用的部分修正：")
mB_bal = fit_logit(sB["train"], dechow_vars, class_weight="balanced")
predB_bal = mB_bal.predict_proba(sB["test"][dechow_vars].values)[:, 1]
metB_bal = eval_metrics(predB_bal, sB["test"]["misstate"].values)
print(f"  Model B + balanced: AUC={metB_bal['AUC']:.4f}  "
      f"NDCG@100={metB_bal['NDCG']:.4f}  Recall@1%={metB_bal['RecallP']:.4f}")

print("\n[A] AUC=%.4f NDCG=%.4f R@1%%=%.4f P@1%%=%.4f"
      % (metA['AUC'], metA['NDCG'], metA['RecallP'], metA['PrecP']))
print("[B] AUC=%.4f NDCG=%.4f R@1%%=%.4f P@1%%=%.4f"
      % (metB['AUC'], metB['NDCG'], metB['RecallP'], metB['PrecP']))
print("Done.")
